### ISSUE TITLE
Design decision for hybrid config/view-functors.

### ISSUE THREAD

**Issue by rrahn** [seqan/seqan3#286]
So at the moment we have mostly functor implementations that deal with runtime parameters provided by the `(...)` functor call.
It would be also easy to provide compile time values by using `<...>` brackets.
So consider the example of setting free_end_gaps for an alignment configuration:

Runtime version:
```cpp
auto cfg = some_cfg | with_free_end_gaps(free_end_gaps::front_1st | free_end_gaps::back_1st);
```
or as compile time version:
```cpp
auto cfg = some_cfg | with_free_end_gaps<free_end_gaps::front_1st | free_end_gaps::back_1st>;
```

Both implementations would be possible and are desired with respect to the view-design, but cannot co-exists with the same name.

### Design 1
To have both versions at the same time the first example must be adapted to:
```cpp
auto cfg = some_cfg | with_free_end_gaps<>(free_end_gaps::front_1st | free_end
```
adding the empty template-`<>` signature.

### Design 2
Another proposal would be to provide for these special functors a second name with a `_c` suffix:

```cpp
auto cfg = some_cfg | with_free_end_gaps(free_end_gaps::front_1st | free_end_gaps::back_1st)
auto cfg = some_cfg | with_free_end_gaps_c<free_end_gaps::front_1st | free_end_gaps::back_1st>
```

### Design 3
Another solution would be, to provide the functor as a class-wrapper with template deduction, but I would not prefer this solution as the behaviour would differ to much from the current view-design as for the 
compile-time version the default constructor must be invoked.

```cpp
auto cfg = some_cfg | with_free_end_gaps(free_end_gaps::front_1st | free_end_gaps::back_1st)  // OK: template deduction, such that it looks like a functor.
auto cfg = some_cfg | with_free_end_gaps<free_end_gaps::front_1st | free_end_gaps::back_1st>() // Unexpected: need to call the default c-tor to create an instance
```

For example for `view::reverse` we can but we don't need to call the `()`-operator and it is usually not done if the argument list is empty.

Design 1 has the advantage, that for templatized functors with default template parameter the syntax would be the same:
```cpp
auto rng = some_rng | my_view<>(p1, p2, ...)
```
But in the respective case we actually don't want to pass arguments additionally. We either want to pass the arguments as runtime parameters or as compile time parameters and not combining them. 
I think design 2 can address this scenario most properly.
Or I might miss something here?
What do you think?

**Comment by h-2** [comment]
I agree with your overall assessment, although I think it's likely we will get variable template argument deduction guides sometime in the future which would make the original design possible. (Function template argument deduction guides are currently being discussed in http://wg21.link/p1021r0 ).

There is a another possible solution, similar to Design 2:

```cpp
auto cfg = some_cfg | with_free_end_gaps(free_end_gaps::front_1st | free_end_gaps::back_1st)
auto cfg = some_cfg | with_free_end_gaps(c<free_end_gaps::front_1st | free_end_gaps::back_1st>)
```
where c would be
```cpp
template <cfg_enum i>
std::integral_constant<cfg_enum, i> c;
```
It's advantage over Design 2 is that you design a single functor with two `operator()`. This is also good for doxygen, because all our config elements would be functors, and compile-time and run-time would be two different functions of the same entity. In Design 2 it is unclear whether both would be variables (one functor that *returns something*, one variable template that *is* the thing); or whether the first would be a class type that gets constructed and then "is the thing". Both cases will be more difficult to document and code.

The disadvantage is more visual noise and introducing `c` in our global namespace or having to use something longer like `fixed<>`.

**Comment by rrahn** [comment]
Well, for Design 2 I would request to use the same fn_impl with an additional bool template that selects between the runtime or the compile time behavior in the call operator. But I totally agree that documentation wise this is not really optimal. I‘d love to see the proposal above to be added to the language. That would be optimal. Until then I believe your suggestions would be better for the intended usage and Adaption if the proposal get‘s in

**Comment by marehr** [comment]
So we use Design 4?

I'm not sure if we should call it `c<>`. That's awfully short.

**Comment by rrahn** [comment]
I would vote for it. And call it maybe `as_c`?

**Comment by h-2** [comment]
> I'm not sure if we should call it c<>. That's awfully short.

Yes.

> I would vote for it. And call it maybe as_c?

I am ok with `as_c<>` or `fixed<>`.

Btw, here's a more generic definition:
```cpp
template <auto val>
inline constexpr std::integral_constant<decltype(val), val> as_c;
```

**Comment by marehr** [comment]
```cpp
template <auto val>
inline constexpr std::integral_constant<decltype(val), val> as_c;
```
The main problem is that this (currently) does not allow `std::optional` or other more complex types.

Why not `as_ constexpr`?

**Comment by rrahn** [comment]
Well, the thing is that the `_c` suffix is often used to represent a constant value of native types, i.e. bool, int, enum etc.. 
Second, how are you going to provide a non-type template argument, which is a complex type?
Something like:
```cpp
constexpr std::optional<int> opt{1};
foo(as_c<opt>);
```
cannot work, can it?

**Comment by h-2** [comment]
> The main problem is that this (currently) does not allow std::optional or other more complex types.

But if you can't currently pass it as a template parameter you won't be able to pass a value constexpr anyway, in any of the designs. Also I don't know why would want to have complex datatypes here, this is configuration!

> Why not as_ constexpr?

I think people should not have to know what constexpr is to be able to use this interface (although of course it is helpful).

**Comment by smehringer** [comment]
@rrahn Whats the status here? I believe all design decisions regarding the config are done so can this be closed?
